synapse/handlers/sync.py

```diff
@@ -135,6 +138,12 @@ def wait_for_sync_for_user(self, sync_config, since_token=None, timeout=0,
             A Deferred SyncResult.
         """
 
+        if sync_config.is_guest:
+            for room_id in sync_config.filter.list_rooms():
+                world_readable = yield self._is_world_readable(room_id)
+                if not world_readable:
+                    raise AuthError(403, "Guest access not allowed")
```

Comment:
This is an interesting one because if I'm reading this right this means that if you give just a single bad room ID the entire `/sync` will fail. Clients aren't fed enough information to resolve this programatically either. I'd like to see some keys to say "these room IDs are bad" so I can remove them programatically; the alternative (silently dropping bad IDs) leaves a bad taste in my mouth.